### DISCUSSION TITLE
Something odd with rangeStrategy and cargo

### DISCUSSION THREAD

**Discussion by rbtcollins** [renovatebot/renovate#19744] [Suggest an Idea]
So reading the docs, what I expect the rangeStrategy default of 'Replace' to do for a change like this:
Bumps [tokio](https://github.com/tokio-rs/tokio) from 1.21.2 to 1.23.1.

And a dependency like:
```
[dependencies.tokio]
features = ["macros", "net", "rt-multi-thread"]
version = "1.21.2"
```

is to do a lockfile only change. But its rewriting `Cargo.toml`.

Why do I have this expectation? 
https://docs.renovatebot.com/configuration-options/#rangestrategy says the default is "replace".
Replace is documented as "replace = Replace the range with a newer one if the new version falls outside it, and update nothing otherwise"
And per https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html "1.12.2" is not an exact pinned version, but a SemVer range: 
```
The string "0.1.12" is a version requirement. Although it looks like a specific version of the time crate, it actually specifies a range of versions and allows [SemVer](https://semver.org/) compatible updates. An update is allowed if the new version number does not modify the left-most non-zero digit in the major, minor, patch grouping. In this case, if we ran cargo update -p time, cargo should update us to version 0.1.13 if it is the latest 0.1.z release, but would not update us to 0.2.0. If instead we had specified the version string as 1.0, cargo should update to 1.1 if it is the latest 1.y release, but not 2.0. The version 0.0.x is not considered compatible with any other version.

Here are some more examples of version requirements and the versions that would be allowed with them:

1.2.3  :=  >=1.2.3, <2.0.0
1.2    :=  >=1.2.0, <2.0.0
1      :=  >=1.0.0, <2.0.0
0.2.3  :=  >=0.2.3, <0.3.0
0.2    :=  >=0.2.0, <0.3.0
0.0.3  :=  >=0.0.3, <0.0.4
0.0    :=  >=0.0.0, <0.1.0
0      :=  >=0.0.0, <1.0.0
```
So the old tokio version should allow >=1.21.2, <2.0.0 which replace should then not touch.

Is this right? Should I file a bug? Is it shallow to fix?

**Comment by rarkins** [comment]
I haven't looked into the source code but my understanding matches yours - 1.23.1 should have matched the existing range and nothing updated

**Reply by rbtcollins** [reply]
Ok, well we see this on all our repos, and until recently I'd ignored it as a misconfiguration, but then I went to fix it and found we were using the default.... when I search for rangeStrategy in our repo logs nothing is found.

**Reply by rarkins** [reply]
Can you create a minimal reproduction?

**Reply by rbtcollins** [reply]
Yes. 

PR changing Cargo.toml incorrectly: https://github.com/cognitedata/renovate-range-reproducer/pull/1/files

**Reply by rarkins** [reply]
@Turbo87 do you want to cover this in your planned changes?

**Reply by Turbo87** [reply]
I guess this is related to https://github.com/renovatebot/renovate/issues/19809. I ran out of time unfortunately, but https://github.com/Turbo87/semver-cargo could probably be used to fix this.
